s3store/s3store.go

```diff
@@ -536,3 +549,21 @@ func isAwsError(err error, code string) bool {
 	}
 	return false
 }
+
+func (store S3Store) CalcPartSize(size *int64) (effectivePartSize int64) {
+	switch {
+	case *size > store.MaxPartSize*10000:
+		effectivePartSize = store.MaxPartSize
+	case *size <= store.MinPartSize:
+		effectivePartSize = store.MinPartSize
+	case *size/10000 < store.MinPartSize:
+		effectivePartSize = store.MinPartSize
+	case int64(math.Mod(float64(*size), 10000)) == 0:
+		effectivePartSize = *size / 10000
+	case *size/9999 >= store.MaxPartSize:
+		effectivePartSize = store.MaxPartSize
+	default:
+		effectivePartSize = *size / 9999
+	}
+	return effectivePartSize
```

Comment:
The proposed "algorithm" was rather a rough human language description of what needs to be implemented, than an actual algorithm. The switch statement here does implement that outline and I have tested it thoroughly. The test case does test all relevant corner cases. But I have run the code for all sizes from MaxPartSize*9999-1 upwards.